File: w1/camel_case.py
"""
Camel Case - Challenge 5, Week 1.
https://www.hackerrank.com/challenges/three-month-preparation-kit-camel-case/problem
"""

import sys


# set to true if for hackerrank submission
PRINT=False


def camel_case(token: str) -> str:
    """
    A program that creates or splits Java style Camel Case variable, method, and class names.

    parameters:
        token (str): a semi-colon seperated string containing configuration and a target string
                    C: combine, S: seperate
                    M: methdod, C: class, V: variable

    returns:
        str: a camel case string coverted for variable, method or class name configuration

    examples"
        >>> camel_case('C;V;can of coke')
        'canOfCoke'
        >>> camel_case('S;M;sweetTea()')
        'sweet tea'
        >>> camel_case('C;C;mirror')
        'Mirror'
        >>> camel_case('C;M;santa claus')
        'santaClaus()'
    """
    # split into operation [C|S], variant[M|C|V] and string
    operation, variant, target = token.split(';')

    # C is to combine
    # In the case of a combine operation, the words will be a space-delimited
    # list of words starting with lowercase letters that you need to combine
    # into the appropriate camel case String. Methods should end with an empty
    # set of parentheses to differentiate them from variable names.
    if operation == 'C':

        # split on spaces and rejoin as TitleCase
        result = ''.join([t.title() for t in target.split()])

        # v is a variable, so lowercase first item
        if variant == 'V':
            result = result[0].lower() + result[1:]
        
        # m is method, so we need parenthesis and a lowercase start
        if variant == 'M':
            if result[-2:] != '()':
                result = f'{result}()'
                result = result[0].lower() + result[1:]

    # S is to split
    # in the case of a split operation, the words will be a camel case method,
    # class or variable name that you need to split into a space-delimited list
    # of words starting with a lowercase letter.
    elif operation == 'S':

        # split on uppercase letters and add spaces
        result = ''.join(
            map(lambda x: ' ' + x.lower() if x.isupper() else x, target)
        )

        # if first character is a space, remove it
        if result[0] == ' ':
            result = result[1:]

        # M is a method, so ensure we have parenthesis
        if result[-2:] == '()':
            result = result.replace('()', '')
    
    # if we're on HackerRank, print it
    if PRINT:
        print(result)

    # all done, return result
    return result


# HackerRank made you write all of the runner for this one as well.
# that doesn't seem all that 'basic'
if __name__ == '__main__' and PRINT:
    for line in sys.stdin.readlines(): 
        camel_case(line.strip())

File: w1/camel_case.ex
defmodule HackerRank.W1.CamelCase do
@moduledoc """
Camel Case – Challenge 5, Week 1.

[HackerRank Problem – Camel Case](https://www.hackerrank.com/challenges/three-month-preparation-kit-camel-case/problem)
"""

@doc """
Creates or splits Java-style CamelCase variable, method, and class names.

## Parameters

  - `token` (string): A semicolon-separated string specifying the operation and target string.
    Format:
    - `C` = combine, `S` = separate
    - `M` = method, `C` = class, `V` = variable

## Returns

  - A string converted to or from CamelCase based on the specified configuration.

## Examples

    iex> CamelCase.challenge("C;V;can of coke")
    "canOfCoke"

    iex> CamelCase.challenge("S;M;sweetTea()")
    "sweet tea"

    iex> CamelCase.challenge("C;C;mirror")
    "Mirror"

    iex> CamelCase.challenge("C;M;santa claus")
    "santaClaus()"
"""

  @spec challenge(String.t()) :: String.t()
  def challenge(token) do
    # split the token up
    [o, v, t] = String.split(token, ";")

    # pass it off to the specific operation
    case [o, v] do
      ["C", "M"] -> combine_method(t)
      ["C", "V"] -> combine_variable(t)
      ["C", "C"] -> combine_class(t)
      ["S", "M"] -> split_method(t)
      ["S", "V"] -> split_variable(t)
      ["S", "C"] -> split_class(t)
    end
  end

  @doc """
  Combines a string into class name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.combine_class("mirror")
    "Mirror"
  """
  @spec combine_class(String.t()) :: String.t()
  def combine_class(t) do
    t
    |> String.split()
    |> Enum.map(&String.capitalize/1)
    |> Enum.join()
  end

  @doc """
  Combines a string into method name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.combine_method("santa claus")
    "santaClaus()"
  """
  @spec combine_method(String.t()) :: String.t()
  def combine_method(t) do
    # split up the first from the rest
    [first | rest] = String.split(t)

    # downcase the first, capitalise the rest and join it
    name =
      [String.downcase(first) | Enum.map(rest, &String.capitalize/1)]
      |> Enum.join()

    # whack a set of parenthesis on the end
    name <> "()"
  end

  @doc """
  Combines a string into variable name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.combine_variable("can of coke")
    "canOfCoke"
  """
  @spec combine_variable(String.t()) :: String.t()
  def combine_variable(t) do
    # split up the first from the rest
    [first | rest] = String.split(t)

    # downcase the first, capitalise the rest and join it
    [String.downcase(first) | Enum.map(rest, &String.capitalize/1)]
    |> Enum.join()
  end
  @doc """
  Splits a string from a class name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.split_class("sweetTea")
    "sweet tea"
  """
  @spec split_class(String.t()) :: String.t()
  def split_class(t) do
    t
    |> split_camel_case()
    |> Enum.map(&String.downcase/1)
    |> Enum.join(" ")
  end
  @doc """
  Splits a string from method name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.split_method("sweetTea()")
    "sweet tea"
  """
  @spec split_method(String.t()) :: String.t()
  def split_method(t) do
    t
    |> split_camel_case()
    |> Enum.map(&String.downcase/1)
    |> Enum.join(" ")
    |> strip_parens()
  end

  @doc """
  Splits a string from variable name

  ## Parameters
  - `t`: a string to convert

  ## Returns
  - The formatted string

  ## Example

    iex> CamelCase.split_variable("epsonPrinter")
    "epson printer"
  """
  @spec split_variable(String.t()) :: String.t()
  def split_variable(t) do
    t
    |> split_camel_case()
    |> Enum.map(&String.downcase/1)
    |> Enum.join(" ")
  end

  defp split_camel_case(t), do: Regex.scan(~r/[a-z]+|[A-Z][a-z]*/, t) |> List.flatten |> Enum.map(&String.trim/1)
  defp strip_parens(t), do: String.replace(t, "()", "")

end
